### PR TITLE
Support tokenize nested dict in `mars.utils.tokenize`

### LINKED ISSUE #661: [BUG] `mars.utils.tokenize` cannot tokenize nested dict
**Describe the bug**

`mars.utils.tokenize` cannot tokenize nested dict.

**To Reproduce**

```
In [1]: import mars.tensor as mt                                                

In [2]: a = mt.random.rand(1000, 1000, chunk_size={1: 200})                     
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-2-a02b4ff8a2e0> in <module>
----> 1 a = mt.random.rand(1000, 1000, chunk_size={1: 200})

~/Workspace/mars/mars/tensor/random/rand.py in rand(random_state, *dn, **kw)
     83             raise ValueError('rand got unexpected key arguments {0}'.format(key))
     84 
---> 85     return op(chunk_size=chunk_size)

~/Workspace/mars/mars/tensor/random/rand.py in __call__(self, chunk_size)
     31 
     32     def __call__(self, chunk_size=None):
---> 33         return self.new_tensor(None, None, raw_chunk_size=chunk_size)
     34 
     35 

~/Workspace/mars/mars/tensor/operands.py in new_tensor(self, inputs, shape, dtype, order, **kw)
     75             raise TypeError('cannot new tensor with more than 1 outputs')
     76 
---> 77         return self.new_tensors(inputs, shape=shape, dtype=dtype, order=order, **kw)[0]
     78 
     79 

~/Workspace/mars/mars/tensor/operands.py in new_tensors(self, inputs, shape, dtype, order, chunks, nsplits, output_limit, kws, **kw)
     69                     output_limit=None, kws=None, **kw):
     70         return self.new_tileables(inputs, shape=shape, chunks=chunks, nsplits=nsplits,
---> 71                                   output_limit=output_limit, kws=kws, dtype=dtype, order=order, **kw)
     72 
     73     def new_tensor(self, inputs, shape, dtype=None, order=None, **kw):

~/Workspace/mars/mars/core.py in new_tileables(self, inputs, kws, **kw)
    752         """
    753 
--> 754         tileables = self._new_tileables(inputs, kws=kws, **kw)
    755         if is_eager_mode():
    756             ExecutableTuple(tileables).execute(fetch=False)

~/Workspace/mars/mars/tensor/random/core.py in _new_tileables(self, inputs, kws, **kw)
    267         with self._get_inputs_shape_by_given_fields(inputs, shape, raw_chunk_size, True) as (inputs, shape):
    268             kw['shape'] = shape
--> 269             return super(TensorRandomOperandMixin, self)._new_tileables(inputs, kws=kws, **kw)
    270 
    271     def _new_chunks(self, inputs, kws=None, **kw):

~/Workspace/mars/mars/core.py in _new_tileables(self, inputs, kws, **kw)
    726             if kws:
    727                 create_tensor_kw.update(kws[j])
--> 728             tileable = self._create_tileable(j, **create_tensor_kw)
    729             tileables.append(tileable)
    730 

~/Workspace/mars/mars/tensor/operands.py in _create_tileable(self, output_idx, **kw)
     63             kw['_nsplits'] = nsplits
     64         data = TensorData(_shape=shape, _dtype=dt, _order=order,
---> 65                           _op=self, _chunks=chunks, **kw)
     66         return tensor_cls(data)
     67 

~/Workspace/mars/mars/core.py in __init__(self, *args, **kwargs)
    373             kwargs['_nsplits'] = tuple(tuple(s) for s in kwargs['_nsplits'])
    374 
--> 375         super(TileableData, self).__init__(*args, **kwargs)
    376 
    377         if hasattr(self, '_chunks') and self._chunks:

~/Workspace/mars/mars/core.py in __init__(self, *args, **kwargs)
     44 
     45         if self._init_update_key_ and (not hasattr(self, '_key') or not self._key):
---> 46             self._update_key()
     47         if not hasattr(self, '_id') or not self._id:
     48             self._id = str(id(self))

~/Workspace/mars/mars/core.py in _update_key(self)
     68 
     69     def _update_key(self):
---> 70         self._obj_set('_key', tokenize(type(self), *self._values_))
     71         return self
     72 

~/Workspace/mars/mars/_utils.pyx in mars._utils.tokenize()

~/Workspace/mars/mars/_utils.pyx in mars._utils.build_canonical_bytes()

~/Workspace/mars/mars/_utils.pyx in mars._utils.h()

~/Workspace/mars/mars/_utils.pyx in mars._utils.h_iterative()

~/Workspace/mars/mars/_utils.pyx in mars._utils.h_non_iterative()

TypeError: Cannot generate token for {1: 200}, type: <class 'dict'>
```

### PR DESCRIPTION
**Pull request by hekaisheng**
## What do these changes do?

Support tokenize nested dict in `mars.utils.tokenize` and provide class `Tokenizer ` to make code neat and achieve better performance.

## Related issue number
Fixes #661

### CODE REVIEW

File: mars/_utils.pyx [hunk 1]
```diff
@@ -93,57 +93,50 @@ cdef inline to_hex(bytes s):
...
+            handler = self._handlers[object_type]
+            return handler(obj)
+        except KeyError:
+            if hasattr(obj, '_key'):
+                return getattr(obj, '_key')
+            if hasattr(obj, '__mars_tokenize__'):
+                return self.tokenize(obj.__mars_tokenize__())
+
+            for clz in self._handlers.keys():
+                if issubclass(object_type, clz):
```
Comment (wjsi) [inline mars/_utils.pyx]: Maybe iterating over ``__mro__`` will be a faster solution?

Comment (qinxuye) [reply mars/_utils.pyx]: > Maybe iterating over `__mro__` will be a faster solution?

Agreed.

File: mars/_utils.pyx [hunk 2]
```diff
@@ -93,57 +93,50 @@ cdef inline to_hex(bytes s):
...
 
-cdef inline list h_iterative(object ob):
+    cdef inline tokenize(self, object obj):
+        object_type = type(obj)
+        try:
+            handler = self._handlers[object_type]
+            return handler(obj)
+        except KeyError:
+            if hasattr(obj, '_key'):
+                return getattr(obj, '_key')
```
Comment (qinxuye) [inline mars/_utils.pyx]: Just return `obj._key`.

File: mars/_utils.pyx [hunk 3]
```diff
@@ -93,57 +93,50 @@ cdef inline to_hex(bytes s):
...
     nested = deque(ob)
     h_list = []
     dq = deque()
     while nested or dq:
         x = dq.pop() if dq else nested.popleft()
         if isinstance(x, (list, tuple)):
-            dq.extend(reversed(x))
+            dq.extend(x)
+        elif isinstance(x, set):
+            dq.extend(sorted(x, key=str))
```
Comment (qinxuye) [inline mars/_utils.pyx]: Is it possible that just `sorted(x)`?

File: mars/_utils.pyx [hunk 4]
```diff
@@ -183,25 +176,50 @@ cdef inline _extract_range_index_attr(object range_index, str attr):
...
+    tokenize_handler.register(t, lambda ob: ob)
+
+for t in (np.dtype, np.generic):
+    tokenize_handler.register(t, lambda ob: repr(ob))
+
+for t in (list, tuple):
+    tokenize_handler.register(t, iterative_tokenize)
 
+tokenize_handler.register(np.ndarray, tokenize_numpy)
+tokenize_handler.register(dict, lambda ob: iterative_tokenize(sorted(list(ob.items()), key=str)))
```
Comment (qinxuye) [inline mars/_utils.pyx]: Since `iterative_tokenize` can handle dict, is it duplicated to define the same logic here?

File: mars/_utils.pyx [hunk 5]
```diff
@@ -183,25 +176,50 @@ cdef inline _extract_range_index_attr(object range_index, str attr):
...
+tokenize_handler.register(set, lambda ob: iterative_tokenize(sorted(ob, key=str)))
```
Comment (qinxuye) [inline mars/_utils.pyx]: Ditto.

File: mars/_utils.pyx (hunk 2 above)
Comment (hekaisheng) [reply mars/_utils.pyx]: Just to avoid access to a protected member, will fix

File: mars/_utils.pyx (hunk 3 above)
Comment (hekaisheng) [reply mars/_utils.pyx]: I think it is okay.

File: mars/_utils.pyx (hunk 4 above)
Comment (hekaisheng) [reply mars/_utils.pyx]: Should be sorted before call `iterative_tokenize `.

**Comment by hekaisheng** [comment]
I generated a big graph and ran a rough benchmark as below.

``` Python
In [1]: import numpy as np 
   ...: import mars.tensor as mt 
   ...:  
   ...: a = mt.random.random((100, 100), chunk_size=10) 
   ...: for _ in range(40): 
   ...:     temp = mt.random.rand(100, 100, chunk_size=10) 
   ...:     a = (a + temp).dot(a) 
   ...: graph = a.build_graph(compose=True, tiled=True)                                                                             

In [2]: from mars.utils import tokenize                                                                                             

In [3]: %timeit _ = [tokenize(type(node), *(getattr(node, k, None) for k in node.__slots__)) for node in graph.iter_nodes()]        
683 ms ± 7.3 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [3]: %timeit _ = [tokenize2(type(node), *(getattr(node, k, None) for k in node.__slots__)) for node in graph.iter_nodes()]        
742 ms ± 22.2 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```
`tokenize2` is the old version, improved by about 8%.

**Comment by qinxuye** [comment]
> I generated a big graph and ran a rough benchmark as below.
> 
> ```python
> In [1]: import numpy as np 
>    ...: import mars.tensor as mt 
>    ...:  
>    ...: a = mt.random.random((100, 100), chunk_size=10) 
>    ...: for _ in range(40): 
>    ...:     temp = mt.random.rand(100, 100, chunk_size=10) 
>    ...:     a = (a + temp).dot(a) 
>    ...: graph = a.build_graph(compose=True, tiled=True)                                                                             
> 
> In [2]: from mars.utils import tokenize                                                                                             
> 
> In [3]: %timeit _ = [tokenize(type(node), *(getattr(node, k, None) for k in node.__slots__)) for node in graph.iter_nodes()]        
> 683 ms ± 7.3 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
> 
> In [3]: %timeit _ = [tokenize2(type(node), *(getattr(node, k, None) for k in node.__slots__)) for node in graph.iter_nodes()]        
> 742 ms ± 22.2 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
> ```
> 
> `tokenize2` is the old version, improved by about 8%.

Nice job, the consequence is quite acceptable.

**Pull request merged**